mwtext/wikitext_preprocessor_mwpfh.py

```diff
@@ -0,0 +1,290 @@
+import mwparserfromhell
+from mwparserfromhell.nodes import ExternalLink, Heading, Tag, Text, Wikilink
+
+
+logger = logging.getLogger(__name__)
+
+
+FORBIDDEN_SECTIONS = [
+    "bibliography",
+    "citations",
+    "external links",
+    "further reading",
+    "other uses",
+    "references",
+    "see also",
+    "sources",
+]
+
+FORBIDDEN_WIKILINK_PREFIXES = [
+    "file:",
+    "image:",
+    "category:",
+]
+
+ALLOWED_TAGS = [
+    "b",
+    "i",
+    "u",
+]
+
+WikilinkParser = Callable[[Wikilink], Tuple[bool, bool, str, str]]
+
+
+class WikitextPreprocessorMwpfh:
+
+    """Wikitext Preprocessor using MediaWiki Parser From Hell
+
+    Args:
+        forbidden_wikilink_prefixes (Iterable[str]): ignore wikilinks with these
+            (case insensitive) prefixes.
+        forbidden_sections (Iterable[str]): skip sections with these
+            (case insensitive) titles.
+        allowed_tags (Iterable[str]): include the contents of these tags
+        include_disallowed_tag_tokens (bool): if True, include a single
+            token for tags that are not in the `allowed_tags` list.  For
+            example a <table> token to represent all the markup of a table.
+        canonicalize_wikilink_targets (bool): if True, uppercase the first character
+            of link targets and replace spaces with underscores (to match titles
+            in the SQL dumps)
+        custom_wikilink_parser (Callable[[Wikilink], Tuple[bool, bool, str, str]]):
+            function that takes in a Wikilink object and returns a 4-tuple
+            (add_link, add_text, target, anchor) where,
+              * add_link: determines if link is added to link list
+              * add_text: determines if anchor is added to text stream
+              * target: target page title to use
+              * anchor: anchor text to use
+            Uses _default_wikilink_parser if None.
+    """
+    def __init__(
+        self,
+        forbidden_wikilink_prefixes: Iterable[str] = FORBIDDEN_WIKILINK_PREFIXES,
+        forbidden_sections: Iterable[str] = FORBIDDEN_SECTIONS,
+        allowed_tags: Iterable[str] = ALLOWED_TAGS,
+        include_disallowed_tag_tokens: bool = False,
+        canonicalize_wikilink_targets: bool = False,
+        custom_wikilink_parser: Optional[WikilinkParser] = None,
+    ) -> None:
+        self.forbidden_wikilink_prefixes = forbidden_wikilink_prefixes
+        self.forbidden_sections = forbidden_sections
+        self.allowed_tags = allowed_tags
+        self.include_disallowed_tag_tokens = include_disallowed_tag_tokens
+        self.canonicalize_wikilink_targets = canonicalize_wikilink_targets
+        self.custom_wikilink_parser = custom_wikilink_parser
+        self._reset()
+
+    def _reset(self) -> None:
+        self._section_idx = 0
+        self._section_name = "Introduction"
+        self._skipping_section = False
+        self._current_text = ""
+        self._current_wikilinks = []
```

Comment:
If we're resetting things for every parsing job, it might make more sense to construct a new object for each text we're processing.

Reply:
Definitely thought about that.  Will give that a go. All of the parsing state is wrapped up in these variables so couldn't hurt to make this its own class or object like `ParseState` or something.

Reply:
The section logic also assumes that a full page of wikitext is being passed in ... I can make that more clear.